### ISSUE TITLE
Use vector for BlockList in MarkSweepSpace

### ISSUE THREAD

**Issue by wks** [mmtk/mmtk-core#1153]
TL;DR: `MarkSweepSpace` organizes `Block` instances using linked list, which is not as efficient as `Vec`.  As a result, we also sweep blocks differently from Immix, although the sweeping algorithms are similar.  We should test if the linked lists are a performance bottleneck and switch to `Vec` and do things similar to the `ImixSpace`. 

Currently the MiMalloc-inspired `MarkSweepSpace` uses the linked list data structure to organize blocks in a `BlockList`.  Linked lists are not as efficient as `Vec` because we need to visit one node before we know the address of the other.  It has several consequences:

1.  To parallelize the sweeping of blocks, we need to break large `BlockList` into smaller lists.  To do this, we need to traverse the linked list, and load from every block scattered all over the memory.  In comparison, to break a `Vec`, we only need to access the `Vec` itself.  We don't need to access the memory of any `Block` inside the `Vec`.
2.  We cannot use `SweepChunk` like `ImmixSpace` because if we do so, two GC workers will try to remove two blocks from one linked list concurrently, and that's a data race.  We have to let each GC worker sweep blocks from a different `BlockList`, and that means it will access `Block` instances that are not spatially adjacent, which means not friendly to the cache.

The `ImmixSpace` also organizes the space into blocks.  But `ImmixSpace` does not use linked lists, and it is not slow.  `ImmixSpace` does not have mutator-local caches.  Instead, `ImmixSpace` has a global `ReusableBlockPool` which has a lock-free fast path for popping elements from the `BlockQueue`.  The `BlockQueue` is backed by a fixed-sized array, and the `ReusableBlockPool` contains a `Vec` of multiple `BlockQueue` instances.

Should we proiritize this?  Maybe.  *Currently, the sweeping of blocks is still only a small portion of the STW time.*  Meanwhile, *the allocation is slow because we haven't implemented an allocation fast path for OpenJDK*.  Even after we implemented the allocation fast path, it will still be slower than bump-pointer allocation.  So optimizing the linked list data structure may not be immediately profitable.
